src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,478 @@
+    }
+
+    if (head.tail.offset == head.queue_size) {
+      head.tail.offset = head.max_head_size;
+      head.tail.gen += 1;
+    }
+    CLS_LOG(20, "INFO: queue_enqueue: New tail offset: %s \n", head.tail.to_str().c_str());
+  } //end - for
+
+  return 0;
+}
+
+int queue_list_entries(cls_method_context_t hctx, const cls_queue_list_op& op, cls_queue_list_ret& op_ret, cls_queue_head& head)
+{
+  // If queue is empty, return from here
+  if ((head.front.offset == head.tail.offset) && (head.front.gen == head.tail.gen)) {
+    CLS_LOG(20, "INFO: queue_list_entries(): Next offset is %s\n", head.front.to_str().c_str());
+    op_ret.next_marker = head.front.to_str();
+    op_ret.is_truncated = false;
+    return 0;
+  }
+
+  cls_queue_marker start_marker;
+  start_marker.from_str(op.start_marker.c_str());
+  cls_queue_marker next_marker = {0, 0};
+
+  uint64_t start_offset = 0, gen = 0;
+  if (start_marker.offset == 0) {
+    start_offset = head.front.offset;
+    gen = head.front.gen;
+  } else {
+    start_offset = start_marker.offset;
+    gen = start_marker.gen;
+  }
+
+  op_ret.is_truncated = true;
+  uint64_t chunk_size = 1024;
+  uint64_t contiguous_data_size = 0, size_to_read = 0;
+  bool wrap_around = false;
+
+  //Calculate length of contiguous data to be read depending on front, tail and start offset
+  if (head.tail.offset > head.front.offset) {
+    contiguous_data_size = head.tail.offset - start_offset;
+  } else if (head.front.offset >= head.tail.offset) {
+    if (start_offset >= head.front.offset) {
+      contiguous_data_size = head.queue_size - start_offset;
+      wrap_around = true;
+    } else if (start_offset <= head.tail.offset) {
+      contiguous_data_size = head.tail.offset - start_offset;
+    }
+  }
+
+  CLS_LOG(10, "INFO: queue_list_entries(): front is: %s, tail is %s\n", head.front.to_str().c_str(), head.tail.to_str().c_str());
+
+  bool offset_populated = false, entry_start_processed = false;
+  uint64_t data_size = 0, num_ops = 0;
+  uint16_t entry_start = 0;
+  bufferlist bl;
+  do
+  {
+    CLS_LOG(10, "INFO: queue_list_entries(): start_offset is %lu\n", start_offset);
+  
+    bufferlist bl_chunk;
+    //Read chunk size at a time, if it is less than contiguous data size, else read contiguous data size
+    if (contiguous_data_size > chunk_size) {
+      size_to_read = chunk_size;
+    } else {
+      size_to_read = contiguous_data_size;
+    }
+    CLS_LOG(10, "INFO: queue_list_entries(): size_to_read is %lu\n", size_to_read);
+    if (size_to_read == 0) {
+      next_marker = head.tail;
+      op_ret.is_truncated = false;
+      CLS_LOG(20, "INFO: queue_list_entries(): size_to_read is 0, hence breaking out!\n");
+      break;
+    }
+
+    auto ret = cls_cxx_read(hctx, start_offset, size_to_read, &bl_chunk);
+    if (ret < 0) {
+      return ret;
+    }
+
+    //If there is leftover data from previous iteration, append new data to leftover data
+    bl.claim_append(bl_chunk);
+    bl_chunk = bl;
+    bl.clear();
```

Comment:
```diff
-    bl_chunk = bl;
-    bl.clear();
+    bl_chunk = std::move(bl);
```
the copy assignment will increment the refcount of all segments, and clear() will decrement them again. using move here avoids any ref counting

Reply:
Ok